<div style="border:solid Chocolate 2px; padding: 40px">
Люда, привет! Исследования по кофейням и презентации так и не наблюдаю. В чем проблема? Ты мне не исправленный проект второй раз отправляешь.
  

 <div class="alert alert-info"><b>добрый день! Исследование по кофейням Шаг4. ССылка на презентацию не сохранилась. Исправила.</b> </div>

<div class="alert alert-success">
<h2> Комментарий ревьюера v3<a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Теперь все нашла и оценила! Работа принята, хорошего дня:)
</div>

<div style="border:solid Chocolate 2px; padding: 40px">

<b>Людмила, приветствую!</b>

Буду рада с тобой познакомиться. Меня зовут Евгения Маракова и я буду ревьюером твоего проекта. Если ты не против, то предлагаю общаться на "ты":) Если удобнее на "вы", то нет проблем, только скажи об этом.

В процессе работы ты встретишь комментарии <font color='green'>зеленого</font>, <font color='gold'>желтого</font> и <font color='red'>красного</font> цветов. Что они значат, смотри ниже:
    <br/>

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Решение на этом шаге является полностью правильным.
</div>

<br/>

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Можно лучше⚠️:</b> Решение на этом шаге станет еще лучше, если внести небольшие коррективы.
</div>


<br/>
<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>Необходимо исправить❌:</b>
 Решение на этом шаге требует существенной переработки и внесения правок. Напоминаю, что проект не может быть принят с первого раза, если ревью содержит комментарии, рекомендующие доработать шаги.
</div>
    
Увидев мой комментарий, не удаляй его, он будет очень полезен в случае повторной проверки работы😊
    
Свой комментарий помечай так, как показано ниже, тогда нам будет легче вести диалог, а мне выполнить повторную проверку проекта.
    <div class="alert alert-info"> <b>Комментарий студента: </b> </div>
    


# Рынок заведений общественного питания Москвы

Презентация: <https://cloud.mail.ru/public/VWyT/jYDC9Pmjh> 

In [1]:
!pip install nbconvert

In [2]:
!pip install bokeh

In [3]:
!pip install folium

In [4]:
!pip install geopandas

In [5]:
!pip install pyyaml

<br/>
<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>Необходимо исправить❌:</b>
Дан неверный путь к файлу.

<div class="alert alert-success">
<h2> Комментарий ревьюера v3<a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Отличная информативная презентация!
</div>

Описание проекта
Инвесторы из фонда «Shut Up and Take My Money» решили попробовать себя в новой области и открыть заведение общественного питания в Москве. Заказчики ещё не знают, что это будет за место: кафе, ресторан, пиццерия, паб или бар, — и какими будут расположение, меню и цены.
Для начала они просят вас — аналитика — подготовить исследование рынка Москвы, найти интересные особенности и презентовать полученные результаты, которые в будущем помогут в выборе подходящего инвесторам места.
Постарайтесь сделать презентацию информативной и лаконичной. Её структура и оформление сильно влияют на восприятие информации читателями вашего исследования. Выбирать инструменты (matplotlib, seaborn и другие) и типы визуализаций вы можете самостоятельно.
Вам доступен датасет с заведениями общественного питания Москвы, составленный на основе данных сервисов Яндекс Карты и Яндекс Бизнес на лето 2022 года. Информация, размещённая в сервисе Яндекс Бизнес, могла быть добавлена пользователями или найдена в общедоступных источниках. Она носит исключительно справочный характер.


Описание данных


Файл moscow_places.csv:
name — название заведения;
address — адрес заведения;
category — категория заведения, например «кафе», «пиццерия» или «кофейня»;
hours — информация о днях и часах работы;
lat — широта географической точки, в которой находится заведение;
lng — долгота географической точки, в которой находится заведение;
rating — рейтинг заведения по оценкам пользователей в Яндекс Картах (высшая оценка — 5.0);
price — категория цен в заведении, например «средние», «ниже среднего», «выше среднего» и так далее;
avg_bill — строка, которая хранит среднюю стоимость заказа в виде диапазона, например:
«Средний счёт: 1000–1500 ₽»;
«Цена чашки капучино: 130–220 ₽»;
«Цена бокала пива: 400–600 ₽».
и так далее;
middle_avg_bill — число с оценкой среднего чека, которое указано только для значений из столбца avg_bill, начинающихся с подстроки «Средний счёт»:
Если в строке указан ценовой диапазон из двух значений, в столбец войдёт медиана этих двух значений.
Если в строке указано одно число — цена без диапазона, то в столбец войдёт это число.
Если значения нет или оно не начинается с подстроки «Средний счёт», то в столбец ничего не войдёт.
middle_coffee_cup — число с оценкой одной чашки капучино, которое указано только для значений из столбца avg_bill, начинающихся с подстроки «Цена одной чашки капучино»:
Если в строке указан ценовой диапазон из двух значений, в столбец войдёт медиана этих двух значений.
Если в строке указано одно число — цена без диапазона, то в столбец войдёт это число.
Если значения нет или оно не начинается с подстроки «Цена одной чашки капучино», то в столбец ничего не войдёт.
chain — число, выраженное 0 или 1, которое показывает, является ли заведение сетевым (для маленьких сетей могут встречаться ошибки):
0 — заведение не является сетевым
1 — заведение является сетевым
district — административный район, в котором находится заведение, например Центральный административный округ;
seats — количество посадочных мест.

# Шаг 1. Загрузим данные и изучим общую информацию

In [6]:
import pandas as pd
import re
import seaborn as sns
import matplotlib.pyplot as plt
import plotly.express as px
import numpy as np
from plotly import graph_objects as go

In [7]:
data = pd.read_csv('/datasets/moscow_places.csv')
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8406 entries, 0 to 8405
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   name               8406 non-null   object 
 1   category           8406 non-null   object 
 2   address            8406 non-null   object 
 3   district           8406 non-null   object 
 4   hours              7870 non-null   object 
 5   lat                8406 non-null   float64
 6   lng                8406 non-null   float64
 7   rating             8406 non-null   float64
 8   price              3315 non-null   object 
 9   avg_bill           3816 non-null   object 
 10  middle_avg_bill    3149 non-null   float64
 11  middle_coffee_cup  535 non-null    float64
 12  chain              8406 non-null   int64  
 13  seats              4795 non-null   float64
dtypes: float64(6), int64(1), object(7)
memory usage: 919.5+ KB


In [8]:
data

name  category  \
0                   WoWфли      кафе   
1           Четыре комнаты  ресторан   
2                    Хазри      кафе   
3     Dormouse Coffee Shop   кофейня   
4                Иль Марко  пиццерия   
...                    ...       ...   
8401            Суши Мания      кафе   
8402             Миславнес      кафе   
8403               Самовар      кафе   
8404          Чайхана Sabr      кафе   
8405            Kebab Time      кафе   

                                         address  \
0                     Москва, улица Дыбенко, 7/1   
1             Москва, улица Дыбенко, 36, корп. 1   
2                 Москва, Клязьминская улица, 15   
3            Москва, улица Маршала Федоренко, 12   
4                Москва, Правобережная улица, 1Б   
...                                          ...   
8401               Москва, Профсоюзная улица, 56   
8402  Москва, Пролетарский проспект, 19, корп. 1   
8403      Москва, Люблинская улица, 112А, стр. 1   
8404      Москва, Люблинская улица, 112А, стр. 1   
8405              Москва, Россошанский проезд, 6   

                                  district  \
0          Северный административный округ   
1          Северный административный округ   
2          Северный административный округ   
3          Северный административный округ   
4          Северный административный округ   
...                                    ...   
8401   Юго-Западный административный округ   
8402          Южный административный округ   
8403  Юго-Восточный административный округ   
8404  Юго-Восточный административный округ   
8405          Южный административный округ   

                                                  hours        lat        lng  \
0                                ежедневно, 10:00–22:00  55.878494  37.478860   
1                                ежедневно, 10:00–22:00  55.875801  37.484479   
2     пн-чт 11:00–02:00; пт,сб 11:00–05:00; вс 11:00...  55.889146  37.525901   
3                                ежедневно, 09:00–22:00  55.881608  37.488860   
4                                ежедневно, 10:00–22:00  55.881166  37.449357   
...                                                 ...        ...        ...   
8401                             ежедневно, 09:00–02:00  55.670021  37.552480   
8402                             ежедневно, 08:00–22:00  55.640875  37.656553   
8403                           ежедневно, круглосуточно  55.648859  37.743219   
8404                           ежедневно, круглосуточно  55.648849  37.743222   
8405                           ежедневно, круглосуточно  55.598229  37.604702   

      rating          price                       avg_bill  middle_avg_bill  \
0        5.0            NaN                            NaN              NaN   
1        4.5  выше среднего       Средний счёт:1500–1600 ₽           1550.0   
2        4.6        средние         Средний счёт:от 1000 ₽           1000.0   
3        5.0            NaN  Цена чашки капучино:155–185 ₽              NaN   
4        5.0        средние         Средний счёт:400–600 ₽            500.0   
...      ...            ...                            ...              ...   
8401     4.4            NaN                            NaN              NaN   
8402     4.8            NaN                            NaN              NaN   
8403     3.9            NaN          Средний счёт:от 150 ₽            150.0   
8404     4.2            NaN                            NaN              NaN   
8405     3.9            NaN                            NaN              NaN   

      middle_coffee_cup  chain  seats  
0                   NaN      0    NaN  
1                   NaN      0    4.0  
2                   NaN      0   45.0  
3                 170.0      0    NaN  
4                   NaN      1  148.0  
...                 ...    ...    ...  
8401                NaN      0   86.0  
8402                NaN      0  150.0  
8403                NaN      0  150.0  
8404                NaN      1  150.0  
8

Всего 8406 заведений. 

Всего 8406 заведений представлено. 

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Молодец, ты отлично познакомилась с данными!
</div>

# Шаг 2. Выполним предобработку данных

Проверим данные на пропуски и дубликаты.

In [9]:
data.isnull().sum()

name                    0
category                0
address                 0
district                0
hours                 536
lat                     0
lng                     0
rating                  0
price                5091
avg_bill             4590
middle_avg_bill      5257
middle_coffee_cup    7871
chain                   0
seats                3611
dtype: int64

В данных есть пропуски. Пропущены данные в столбце hours(информация о днях и часах работы заведения). Скорее всего либо нет данных или данные искажены, либо кафе уже не работает. Также нет информации в ценном сегменте и в количесте посадочных мест. У больше половины данных пропущены цены. Оставим пропуски неизменными. 

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Верное решение! Удаление или заполнение пропусков исказит исследование.
</div>

In [10]:
data.duplicated().sum()

0

Дубликатов нет.

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Датасет на наличие явных дублей проверен верно!
</div>

Проверим на неявные дубликаты с учётом регистра.

In [11]:
data['address'] = data['address'].str.lower()
data['name']  = data['name'].str.lower()

In [12]:
data.duplicated().sum()

0

Неявных дубликатов нет.

<br/>
<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>Необходимо исправить❌:</b>
Необходимо также рассмотреть неявные дубли.  Неявные дубликаты могут появиться в случаях, когда данные собирались из разных источников или в системе были какие-то сбои при логировании.
</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера v2<a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Проверка выполнена!
</div>

In [13]:
words = ['улица','ул','переулок','шоссе','проспект','площадь','проезд',
         'село','аллея','бульвар','набережная','тупик','линия']

str_pat = r".*,\s*\b([^,]*?(?:{})\b[^,]*)[,$]+".format("|".join(words))
data['street'] = data['address'].str.extract(str_pat, flags=re.I)


In [14]:
data['is_24/7'] = data['hours']. str.contains('ежедневно, круглосуточно')

In [15]:
data.head(15)

name         category  \
0                 wowфли             кафе   
1         четыре комнаты         ресторан   
2                  хазри             кафе   
3   dormouse coffee shop          кофейня   
4              иль марко         пиццерия   
5           sergio pizza         пиццерия   
6            огни города          бар,паб   
7             mr. уголёк  быстрое питание   
8            donna maria         ресторан   
9                 готика             кафе   
10        great room bar          бар,паб   
11           шашлык шефф             кафе   
12              заправка             кафе   
13               буханка         булочная   
14              у сильвы          бар,паб   

                                    address                         district  \
0                москва, улица дыбенко, 7/1  Северный административный округ   
1        москва, улица дыбенко, 36, корп. 1  Северный административный округ   
2            москва, клязьминская улица, 15  Северный административный округ   
3       москва, улица маршала федоренко, 12  Северный административный округ   
4           москва, правобережная улица, 1б  Северный административный округ   
5              москва, ижорская улица, вл8б  Северный административный округ   
6     москва, клязьминская улица, 9, стр. 3  Северный административный округ   
7     москва, клязьминская улица, 9, стр. 3  Северный административный округ   
8   москва, дмитровское шоссе, 107, корп. 4  Северный административный округ   
9               москва, ангарская улица, 39  Северный административный округ   
10           москва, левобережная улица, 12  Северный административный округ   
11    москва, улица маршала федоренко, 10с1  Северный административный округ   
12           москва, мкад, 80-й километр, 1  Северный административный округ   
13     москва, базовская улица, 15, корп. 1  Северный административный округ   
14            москва, ангарская улица, 42с1  Северный административный округ   

                                                hours        lat        lng  \
0                              ежедневно, 10:00–22:00  55.878494  37.478860   
1                              ежедневно, 10:00–22:00  55.875801  37.484479   
2   пн-чт 11:00–02:00; пт,сб 11:00–05:00; вс 11:00...  55.889146  37.525901   
3                              ежедневно, 09:00–22:00  55.881608  37.488860   
4                              ежедневно, 10:00–22:00  55.881166  37.449357   
5                              ежедневно, 10:00–23:00  55.888010  37.509573   
6                   пн 15:00–04:00; вт-вс 15:00–05:00  55.890752  37.524653   
7   пн-чт 10:00–22:00; пт,сб 10:00–23:00; вс 10:00...  55.890636  37.524303   
8                              ежедневно, 10:00–22:00  55.880045  37.539006   
9                              ежедневно, 12:00–00:00  55.879038  37.524487   
10                           ежедневно, круглосуточно  55.877832  37.469171   
11                             ежедневно, 10:00–21:00  55.881770  37.492362   
12                                  вт-сб 09:00–18:00  55.899938  37.517958   
13                             ежедневно, 08:00–22:00  55.877007  37.504980   
14                             ежедневно, 13:00–00:00  55.885528  37.528371   

    rating          price                       avg_bill  middle_avg_bill  \
0      5.0            NaN                            NaN              NaN   
1      4.5  выше среднего       Средний счёт:1500–1600 ₽           1550.0   
2      4.6        средние         Средний счёт:от 1000 ₽           1000.0   
3      5.0            NaN  Цена чашки капучино:155–185 ₽              NaN   
4      5.0        средние         Средний счёт:400–600 ₽            500.0   
5      4.6        средние                            NaN              NaN   
6      4.4        средние             Средний счёт:199 ₽            199.0   
7      4.7        средние         Средний счёт:200–300 ₽            250.0   
8      4.8        средние          Средний счёт:от 500 ₽    

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Можно лучше⚠️:</b> Стоит вывести таблицу с добавленными столбцами, чтобы убедиться в корректности проделанной работы:)
</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера v2<a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Умничка!
</div>

# Шаг 3. Анализ данных

In [16]:
data['category'].unique()

array(['кафе', 'ресторан', 'кофейня', 'пиццерия', 'бар,паб',
       'быстрое питание', 'булочная', 'столовая'], dtype=object)

'кафе', 'ресторан', 'кофейня', 'пиццерия', 'бар,паб' и т.д. - категории заведений представлены в данных. 

In [ ]:
category = data.groupby('category').agg({'category':'count'})
category.columns ={'count'}
fig = px.bar(category.sort_values(by ='count',ascending=False)).update_xaxes(tickangle=45)
fig.update_layout(
    title='Соотношение видов объектов общественного питания по количеству',
    xaxis_title="Тип объекта",
    yaxis_title="Количество")

В Москве больше всего кафе, ресторанов и кофеинь. 

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Задание выполнено верно!
</div>

In [ ]:
m = data.pivot_table(index ='category',values='seats',aggfunc='median')
m.columns={'mean_seat'}
m['mean_seat'] = round(m['mean_seat'])
fig =px.bar(m.sort_values(by ='mean_seat',ascending=False)).update_xaxes(tickangle=45)
fig.update_layout(
    title='Соотношение видов объектов общественного питания по количеству мест ',
    xaxis_title="Тип объекта",
    yaxis_title="Количество")

Лидируют рестораны, бары,пабы,кофейни и столовые по количеству мест.

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Молодец, что использовала именно медиану, в данных имеются приличные выбросы!
</div>

In [ ]:
fig = go.Figure(data=[go.Pie(labels=data['chain'].value_counts().reset_index()['index'], values = data['chain'].value_counts())])
fig.update_layout(
    title='Соотношение сетевых и несетевых заведений в датасете')
fig.show()

Из диаграммы видно,что сетевых всего 38.1%.Что занимает треть заведений в датасете.

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Круговая диаграмма для данного параметра хорошее решение, даже лучше, чем столбчатая. Молодец, что визуализировала доли, а не суммарное количество!

In [ ]:
dt = data.query('chain ==1')
fig = go.Figure(data=[go.Pie(labels=dt['category'].value_counts().reset_index()['index'], values = dt['category'].value_counts())])
fig.update_layout(
    title='Соотношение сетевых заведений по категориям в датасете')
fig.show()

In [ ]:
df = data.pivot_table(index = 'category',columns ='chain',values ='name',aggfunc ='count')
df.columns = {'n_chain','chain'}
df = df.sort_values(by ='n_chain',ascending =False)
df.columns = {'n_chain','chain'}
fig =px.bar(df).update_xaxes(tickangle=45)
fig.update_layout(
    title='Количество заведений сетевых и несетевых',
    xaxis_title="Название сети",
    yaxis_title="Количество")

In [ ]:
df['share'] =df['chain']/(df['chain']+df['n_chain'])*100
fig =px.bar(df['share'].sort_values(ascending =False)).update_xaxes(tickangle=45)
fig.update_layout(
    title=' категории сетевых заведений, которые преобладают из всех сетевых, относительно общей выборки.  ',
    xaxis_title="Название сети",
    yaxis_title="%")

Чаще всего  булочные, пиццерии и кофейни являются сетевыми.

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Если смотреть по суммарному показателю, то да! Думаю логично, что чаще бывают сетевыми кафе, рестораны и кофейни, так как они преобладают по количеству во всем датасете:)

<div class="alert alert-success">
<h2> Комментарий ревьюера v2<a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Теперь преобладают булочные, пиццерии и кофейни:)
</div>

In [ ]:
name = data.query('chain ==1').groupby(by ='name').agg({'name':'count','seats':'median'})
name.columns ={'count','median'}
name['median'] = round(name['median'])
name = name.sort_values(by='count',ascending =False)
name = name.head(15)
name

In [ ]:
fig =px.bar(name).update_xaxes(tickangle=45)
fig.update_layout(
    title='Количество заведений ТОП-15 сетей  ',
    xaxis_title="Название сети",
    yaxis_title="Количество")

ТОП 15 сетей вошли Шоколадница,Домино'с Пицца, Додо Пицца, One Price Coffee, Яндекс Лавка, Cofix и т.д. Среди них первые 6 довольно известные.
Обычно сетевые заведения характерны большим количеством заведений,но с маленькой посадкой внутри.
Они характерны к категориям заведений:
кафе,
кофейня,
пиццерия,
буханка,
ресторан.

<br/>
<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>Необходимо исправить❌:</b>
Вывод то верный, но что-то с графиком не сходится... Ищем ошибку и исправляем:)

<div class="alert alert-success">
<h2> Комментарий ревьюера v2<a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Молодец!
</div>

Следующие административные районы представлены в датасете:

In [ ]:
data['district'].unique()

In [ ]:
dt = data.pivot_table(index =['district'],columns=['category'],aggfunc={'name': 'count'})
dtt = data.pivot_table(index =['district'],aggfunc={'name': 'count'})
dtt.columns={'total'}
dt = dt.merge(dtt,on ='district')
dt

In [ ]:
fig, ax = plt.subplots(figsize=(20, 15))
sns.heatmap(dt,annot=True,cbar=False)

На тепловой карте видно,что самое большое число завений в Центральном административном округе. Больше всего заведений категории кафе и рестораны во всех районах.Меньше всего завежений в Северо-Западном центральном округе.

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Отлично!
</div>

In [ ]:
fig = px.bar(data.pivot_table(index ='category',values='rating',aggfunc='mean').sort_values(by ='rating',ascending =False))
fig.update_layout(
    title=' Средний рейтинг по категориям заведений ',
    xaxis_title="рейтинг",
    yaxis_title="Категория")

<br/>
<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>Необходимо исправить❌:</b>
Необходимо подписать название графика и отсортировать значения.

<div class="alert alert-success">
<h2> Комментарий ревьюера v2<a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Плюсик
</div>

Средний рейтинг по категориям заведений в среднем больше 4. Не сильно различаются усреднённые рейтинги в разных типах общепита.

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Хорошо!
</div>

In [ ]:
rating_data = data.groupby('district', as_index=False)['rating'].agg('median')
rating_data

In [ ]:
# подключаем модуль для работы с JSON-форматом
import json

# читаем файл и сохраняем в переменной
with open('/datasets/admin_level_geomap.geojson', 'r') as f:
    geo_json = json.load(f)

print(json.dumps(geo_json, indent=2, ensure_ascii=False, sort_keys=True))

In [ ]:
# импортируем карту и хороплет
from folium import Map, Choropleth

# загружаем JSON-файл с границами округов Москвы
state_geo = '/datasets/admin_level_geomap.geojson'
# moscow_lat - широта центра Москвы, moscow_lng - долгота центра Москвы
moscow_lat, moscow_lng = 55.751244, 37.618423

# создаём карту Москвы
m = Map(location=[moscow_lat, moscow_lng], zoom_start=10)

# создаём хороплет с помощью конструктора Choropleth и добавляем его на карту
Choropleth(
    geo_data=state_geo,
    data=rating_data,
    columns=['district', 'rating'],
    key_on='feature.name',
    fill_color='YlGn',
    fill_opacity=0.8,
    legend_name='Медианный рейтинг заведений по районам',
).add_to(m)

# выводим карту
m

В среднем заведения с высокими рейтингами находятся в центре Москвы. 

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Решение на этом шаге является полностью правильным.
</div>

In [ ]:
# импортируем карту и маркер
from folium import Map, Marker
# импортируем кластер
from folium.plugins import MarkerCluster

# moscow_lat - широта центра Москвы, moscow_lng - долгота центра Москвы
moscow_lat, moscow_lng = 55.751244, 37.618423

# создаём карту Москвы
m = Map(location=[moscow_lat, moscow_lng], zoom_start=10)
# создаём пустой кластер, добавляем его на карту
marker_cluster = MarkerCluster().add_to(m)

# пишем функцию, которая принимает строку датафрейма,
# создаёт маркер в текущей точке и добавляет его в кластер marker_cluster
def create_clusters(row):
    Marker(
        [row['lat'], row['lng']],
        popup=f"{row['name']} {row['rating']}",
    ).add_to(marker_cluster)

# применяем функцию create_clusters() к каждой строке датафрейма
data.apply(create_clusters, axis=1)

# выводим карту
m

<br/>
<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>Необходимо исправить❌:</b>
Люда, нам необходимо вывести кластеры, а не маркеры. Для этого нужно добавить MarkerCluster.
</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера v2<a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Все отлично!
</div>

In [ ]:
top_streets = data[data['street'].isin(data.groupby('street')['name'].count().sort_values(ascending = False).head(15).index)]
streets_category = top_streets.groupby(['street','category'],as_index = False)['name'].count()
streets_category.rename({'name':'count'},axis=1, inplace =True)
fig = px.bar(streets_category,x ='street',y ='count',color = 'category')
fig.update_layout(title ='Распредение количества заведений и их категорий по топ-15 улицам',
                 xaxis_title ='Название улицы',
                 yaxis_title='Количество заведений',
                 autosize=False,
                 width=900,
                 height = 550,
                 xaxis ={'categoryorder':'total descending'})
fig.show()

<br/>
<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>Необходимо исправить❌:</b>
Сортировка.
</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера v2<a class="tocSkip"> </h2>

<b>Отлично!✅:</b> +
</div>

Больше всего заведений на улице проспекта Мира. Лидируют категории ресторан, кофейня и кафе. 

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Хорошо!
</div>

In [ ]:
streets = data[data['street'].isin(data.groupby('street')['name'].count().sort_values(ascending = False).index)]
streets_1  = streets.groupby(['street'],as_index = False)['name'].count()
streets_1.rename({'name':'count'},axis=1, inplace =True)
streets_1 = streets_1.query('count ==1')


In [ ]:
street = streets_1.merge(streets,how ='left', on = 'street')
street

Всего 414 улиц с 1 объектом общепита.

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Верно!
</div>

In [ ]:
street.groupby('district', as_index=False)['rating'].agg(['count','median'])

Во всех районах есть улица с одним общественным заведением. Больше всего их в центре, где очень узкие и малеькие улицы.

<br/>
<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>Необходимо исправить❌:</b>
По данным улицам необходимо провести какое-нибудь исследование. Можешь рассмотреть рейтинги или расположение по округам.
</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера v2<a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Принято!
</div>

In [ ]:
data['middle_avg_bill'].isnull().sum()/len(data['name'])*100

В столбце middle_avg_bill пропущено 62%. Сгруппируем данные по районам и заменим пропущенные данные мединным значением.

In [ ]:
for i in data['district'].unique():
    data.loc[(data['district'] == i) &(data['middle_avg_bill'].isna()),'middle_avg_bill'] = data.loc[(data['district'] ==i),'middle_avg_bill'].median()


In [ ]:
data['middle_avg_bill'].isnull().sum()

In [ ]:
district_middle = data.groupby('district')['middle_avg_bill'].median()
district_middle

In [ ]:
# импортируем карту и хороплет
from folium import Map, Choropleth

# загружаем JSON-файл с границами округов Москвы
state_geo = '/datasets/admin_level_geomap.geojson'
# moscow_lat - широта центра Москвы, moscow_lng - долгота центра Москвы
moscow_lat, moscow_lng = 55.751244, 37.618423

# создаём карту Москвы
m = Map(location=[moscow_lat, moscow_lng], zoom_start=10)

# создаём хороплет с помощью конструктора Choropleth и добавляем его на карту
Choropleth(
    geo_data=state_geo,
    data=district_middle,
    columns=['district', 'middle_avg_bill'],
    key_on='feature.name',
    fill_color='YlGn',
    fill_opacity=0.8,
    legend_name='Медианный рейтинг заведений по районам',
).add_to(m)

# выводим карту
m

Самые высокие средние чеки в Центральном административном округе и Западном административном округе. Чем ближе к центре, тем выше средний чек.Средний чек зависит от района, в котором находится общепит. 

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Совершенно верно!
</div>

In [ ]:
data['rating'].unique()

In [ ]:
rating_bad = data.query('rating < 3')
rating_bad

In [ ]:
fig = px.bar(rating_bad.groupby('rating', as_index=False)['middle_avg_bill'].agg(['mean','count']).sort_values(by ='mean',ascending =False))
fig.update_layout(title ='Распредение среденего чека  заведений от рейтинга',
                 xaxis_title ='Рейтинг',
                 yaxis_title='Сумма',
                 autosize=False,
                 width=900,
                 height = 550,
                 xaxis ={'categoryorder':'total descending'})
fig.show()

Чем больше рейтинг, тем больше посещаемость заведения. Тем больше средний чек. Но есть аномальные значения, в которых средний чек в два раза выше,чем в остальных такихже заведений с похожим рейтингом. Либо кафе реально плохое, плюс стоимость блюд выше, либо плохая проходимость, за счет этого высокий 

In [ ]:
fig = px.bar(rating_bad.groupby(['category'], as_index=False)['name'].agg(['count']).sort_values(by='count',ascending =False))
fig.update_layout(title ='Распределение заведений с плохими отзывами по категориям ',
                 xaxis_title ='Название категории ',
                 yaxis_title='Количество заведений',
                 autosize=False,
                 width=900,
                 height = 550,
                 xaxis ={'categoryorder':'total descending'})
fig.show()

Лидируют кафе по количеству плохих отзывов, оно и понятно, так как их больше всего в датасете. Меньше всего в булочных. 

# Шаг 4. Детализируем исследование: открытие кофейни

In [ ]:
cafe = data.query('category == "кофейня"')
cafe['name'].count()

Всего 1413 кофейн в датасете.

<div class="alert alert-success">
<h2> Комментарий ревьюера v3<a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Верно!
</div>

In [ ]:
fig = px.bar(cafe.groupby('district', as_index=False)['name'].agg(['count']).sort_values(by ='count',ascending =False))
fig.update_layout(title ='Распределение кафейн по районам ',
                 xaxis_title ='Название района ',
                 yaxis_title='Количество заведений',
                 autosize=False,
                 width=900,
                 height = 550,
                 xaxis ={'categoryorder':'total descending'})
fig.show()

In [ ]:

c = cafe.groupby('district', as_index=False)['name'].count()
# импортируем карту и хороплет
from folium import Map, Choropleth

# загружаем JSON-файл с границами округов Москвы
state_geo = '/datasets/admin_level_geomap.geojson'
# moscow_lat - широта центра Москвы, moscow_lng - долгота центра Москвы
moscow_lat, moscow_lng = 55.751244, 37.618423

# создаём карту Москвы
m = Map(location=[moscow_lat, moscow_lng], zoom_start=10)

Choropleth(
    geo_data=state_geo,
    data=c,
    columns=['district','name'],
    key_on='feature.name',
    fill_color='YlGn',
    fill_opacity=0.8,
    legend_name='Количество кофейн по районам',
).add_to(m)

# 
m

В центре кофейни преобладают. Чем ближе к центре, тем больше кофейн.

<div class="alert alert-success">
<h2> Комментарий ревьюера v3<a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Отлично!
</div>

In [ ]:
is_24 = cafe[cafe['is_24/7'] ==True]
is_24['name'].count()

Из 1413 кофейн всего 59 кофейн работают круглосуточно. Почти половина из них находятся в центре Москвы.

<div class="alert alert-success">
<h2> Комментарий ревьюера v3<a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Правильно!
</div>

In [ ]:
is_24.groupby('district', as_index=False)['name'].agg(['count'])

In [ ]:
cafe.groupby('rating', as_index=False)['name'].agg(['count']).sort_values(by ='count',ascending =False)

В среднем рейтинг кофейн больше 4. Что показывает хороший результат.

<div class="alert alert-success">
<h2> Комментарий ревьюера v2<a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Согласна!
</div>

In [ ]:
cafe.groupby('district', as_index=False)['rating'].agg(['mean']).sort_values(by ='mean',ascending =False)

Как говорилось ранее чем ближе к центру, тем лучше рейтинг заведения.

In [ ]:
cafe['middle_coffee_cup'].unique()

 Заменим пропущенные значения мединным значением сгруппированных данных по районам

In [ ]:
for i in cafe['district'].unique():
    cafe.loc[(data['district'] == i) &(cafe['middle_coffee_cup'].isna()),'middle_coffee_cup'] = cafe.loc[(cafe['district'] ==i),'middle_coffee_cup'].median()


In [ ]:
cafe.groupby('district')['middle_coffee_cup'].agg(['mean']).sort_values(by ='mean',ascending =False)

В центре чашка кафе обойдутся подороже, чем на окраине города.

<div class="alert alert-success">
<h2> Комментарий ревьюера v3<a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Хорошо!
</div>

# Вывод:

Итак,в датасете 8406 заведений. Из них 'кафе', 'ресторан', 'кофейня', 'пиццерия', 'бар,паб' и т.д. - категории заведений представлены в данных. По количеству заведений по категориям больше всего кафе и ресторанов.
Представлены районы:
'Северный административный округ',
       'Северо-Восточный административный округ',
       'Северо-Западный административный округ',
       'Западный административный округ',
       'Центральный административный округ',
       'Восточный административный округ',
       'Юго-Восточный административный округ',
       'Южный административный округ',
       'Юго-Западный административный округ'.
       
       Больше всего заведений в центре Москвы(Центральный административный округ). Что показывает самый высокий средний чек в Москве.

In [ ]:
data[data['is_24/7'] ==True]

730 заведений работают круглосуточно. 62% заведений не сетевые. В среднем рейтинги заведений во всех районах больше 4. Лидирует Центральный административный округ по среднему рейтингу заведения.

Больше всего заведений на улице проспекта Мира. Лидируют категории ресторан, кофейня и кафе. 

Рекомендация:
    Основателям фонда «Shut Up and Take My Money» лучше всего открыть кафе в  Москвы, на одной из ТОП-15 улиц. Где проходимость наиболее всего и средний чек выше. Также стоит изучить расположение рядом бизнес-центров, университетов и т.п. для лушей проходисмоти и понимания какого сегмента открывать кафе. 

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Отличный вывод!
</div>

<br/>
<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>Необходимо исправить❌:</b>
Не выполнено исследование по кофейням.
</div>

<br/>
<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>Необходимо исправить❌:</b>
Отсутствует презентация.
</div>

<div style="border:solid Chocolate 2px; padding: 40px">
Люда, давай подведем итоги. Ты большая молодец, что старалась и проделала такую большую работу. Мне понравилось, что ты наполнила проект развернутыми выводами и наглядно продемонстрировала каждое задание:)
    
    Разберем над чем стоит поработать:

    1. Делая предобработку данных, не забывай про неявные дубликаты.

    2. Рассмотреть сетевые заведения, относительно общей выборки.

    3. Верно вывести визуализацию "ТОП-15 сетей".

    4. Не забывай про сортировку.

    5. Выполнить исследование по улицам с одним объектом общепита.
    
    6. Выполнить исследование по кофейням.
    
    7. Сделать презентацию.
    
    Пересмотри работу и исправь ошибки. Обязательно пиши, если я могу тебе еще чем-то помочь. Удачи!)
  